examples/petsc/multigrid.c

```diff
@@ -429,24 +428,42 @@ int main(int argc, char **argv) {
     }
   }
 
-  // Setup dummy SNES for AMG coarse solve
-  ierr = SNESCreate(comm, &snes_dummy); CHKERRQ(ierr);
-  ierr = SNESSetDM(snes_dummy, dm[0]); CHKERRQ(ierr);
-  ierr = SNESSetSolution(snes_dummy, X[0]); CHKERRQ(ierr);
-
-  // -- Jacobian matrix
-  ierr = DMSetMatType(dm[0], MATAIJ); CHKERRQ(ierr);
+  // Assemble coarse grid Jacobian for AMG (or other sparse matrix) solve
   ierr = DMCreateMatrix(dm[0], &mat_coarse); CHKERRQ(ierr);
-  ierr = SNESSetJacobian(snes_dummy, mat_coarse, mat_coarse, NULL,
-                         NULL); CHKERRQ(ierr);
-
-  // -- Residual evaluation function
-  ierr = SNESSetFunction(snes_dummy, X[0], FormResidual_Ceed,
-                         user_O[0]); CHKERRQ(ierr);
 
-  // -- Form Jacobian
-  ierr = SNESComputeJacobianDefaultColor(snes_dummy, X[0], mat_O[0],
-                                         mat_coarse, NULL); CHKERRQ(ierr);
+  if (PETSC_VERSION_GE(3,16,0)) { // Assemble matrix analytically
+    CeedInt num_entries, *rows, *cols;
+    CeedVector coo_values;
+    CeedOperatorLinearAssembleSymbolic(user_O[0]->op, &num_entries, &rows, &cols);
+    ISLocalToGlobalMapping ltog_row, ltog_col;
+    ierr = MatGetLocalToGlobalMapping(mat_coarse, &ltog_row, &ltog_col); CHKERRQ(ierr);
+    ierr = ISLocalToGlobalMappingApply(ltog_row, num_entries, rows, rows); CHKERRQ(ierr);
+    ierr = ISLocalToGlobalMappingApply(ltog_col, num_entries, cols, cols); CHKERRQ(ierr);
+    ierr = MatSetPreallocationCOO(mat_coarse, num_entries, rows, cols); CHKERRQ(ierr);
+    free(rows);
+    free(cols);
+    CeedVectorCreate(ceed, num_entries, &coo_values);
+    CeedOperatorLinearAssemble(user_O[0]->op, coo_values);
+    const CeedScalar *values;
+    CeedVectorGetArrayRead(coo_values, CEED_MEM_HOST, &values);
+    ierr = MatSetValuesCOO(mat_coarse, values, ADD_VALUES); CHKERRQ(ierr);
+    CeedVectorRestoreArrayRead(coo_values, &values);
+    CeedVectorDestroy(&coo_values);
+  } else { // Assemble matrix using coloring. Our interfaces match SNES so it's convenient
+    SNES snes_dummy;
+
+    // Setup dummy SNES for AMG coarse solve
+    ierr = SNESCreate(comm, &snes_dummy); CHKERRQ(ierr);
+    ierr = SNESSetDM(snes_dummy, dm[0]); CHKERRQ(ierr);
+    ierr = SNESSetSolution(snes_dummy, X[0]); CHKERRQ(ierr);
+    ierr = SNESSetFunction(snes_dummy, X[0], FormResidual_Ceed,
+                           user_O[0]); CHKERRQ(ierr);
+    ierr = SNESSetJacobian(snes_dummy, mat_coarse, mat_coarse, NULL,
+                           NULL); CHKERRQ(ierr);
+    ierr = SNESComputeJacobianDefaultColor(snes_dummy, X[0], mat_O[0],
+                                           mat_coarse, NULL); CHKERRQ(ierr);
+    ierr = SNESDestroy(&snes_dummy); CHKERRQ(ierr);
+  }
```

Comment:
We should probably leave this fallback in place, at least until after petsc-3.16 is released. (That'll be around the same time or after libceed-0.10, at the end of this month.)

Reply:
For just the multigrid example, I don't see any harm in leaving this there long term. 

We have a lot of old compatibility macros floating around. Do we want to set 3.16 as a line in the sand and require a newer PETSc for fluids and solids?

Reply:
Probably, in that I'd rather not have CI testing multiple older versions of PETSc in the libCEED repository (which libCEED doesn't even depend on). But I envision that we will at some point freeze (or cool) new science/engineering features to these examples and move that to a new repository that can readily use new PETSc features and expose a library interface to the simulation (for use by UQ, etc.).

Reply:
Ah, we were already discussing that here. With #885 I think we can drop this fallback. We already require 3.17 for creating the restrictions.

Reply:
It was helpful for comparison at the time, but I think we can drop it now. This code path isn't interesting to someone learning libCEED. For nonlinear problems, it may still be useful in case the user hasn't written an exact Jacobian, but I'd rather implement finite differencing at quadrature points instead of coloring.